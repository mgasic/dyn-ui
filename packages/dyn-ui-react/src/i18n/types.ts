import type React from 'react';

export type PluralForms = Partial<Record<Intl.LDMLPluralRule | 'other', string>>;

export interface LocaleMessages {
  [key: string]: string | PluralForms | LocaleMessages;
}

export type MessageDescriptor =
  | string
  | {
      id: string;
      defaultMessage?: string;
      values?: Record<string, unknown>;
      count?: number;
      locale?: string;
    };

export interface TranslateOptions {
  count?: number;
  values?: Record<string, unknown>;
  defaultMessage?: string;
  locale?: string;
}

export interface I18nProviderProps {
  /** Initial locale applied to the tree */
  initialLocale?: string;
  /** Controlled locale value */
  locale?: string;
  /** Locale used when no translation is available */
  fallbackLocale?: string;
  /** Dictionary of translations keyed by locale */
  messages?: Record<string, LocaleMessages>;
  /** Detect and apply RTL direction automatically */
  detectDirection?: boolean;
  /** Callback invoked when locale changes */
  onLocaleChange?: (locale: string) => void;
  children: React.ReactNode;
}

export interface I18nContextValue {
  locale: string;
  direction: 'ltr' | 'rtl';
  isRtl: boolean;
  availableLocales: string[];
  messages: LocaleMessages;
  t: (descriptor: MessageDescriptor | null | undefined, options?: TranslateOptions) => string;
  formatNumber: (value: number, options?: Intl.NumberFormatOptions) => string;
  formatDate: (
    value: Date | number | string,
    options?: Intl.DateTimeFormatOptions & { timeZone?: string }
  ) => string;
  setLocale: (locale: string) => void;
  registerMessages: (locale: string, dictionary: LocaleMessages) => void;
}
