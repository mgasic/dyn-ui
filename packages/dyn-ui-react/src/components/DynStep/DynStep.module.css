.root {
  display: flex;
  flex-direction: column;
  gap: var(--dyn-step-gap, var(--dyn-spacing-md, 0.75rem));
  padding: var(--dyn-step-padding, 0);
  width: 100%;
  min-width: 0;
}

.root > * {
  min-width: 0;
}
