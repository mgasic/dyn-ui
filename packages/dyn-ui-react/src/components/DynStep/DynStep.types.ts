import type {
  ComponentPropsWithoutRef,
  ComponentRef,
  CSSProperties,
  ElementType,
} from 'react';
import type { BaseComponentProps } from '../../types';
import type { LayoutSpacing } from '../../types/layout.types';

export type DynStepSpacingValue = LayoutSpacing | number | string;

type PolymorphicComponentProps<E extends ElementType, P> = P &
  Omit<ComponentPropsWithoutRef<E>, keyof P>;

export interface DynStepOwnProps extends BaseComponentProps {
  /** Controls the vertical gap between children */
  spacing?: DynStepSpacingValue;
  /** Optional padding applied to the wrapper */
  padding?: DynStepSpacingValue;
  /** Optional inline style overrides */
  style?: CSSProperties;
}

export type DynStepBaseProps = DynStepOwnProps;

export type DynStepProps<E extends ElementType = 'section'> =
  PolymorphicComponentProps<E, DynStepBaseProps> & {
    as?: E;
  };

export type DynStepRef<E extends ElementType = 'section'> = ComponentRef<E>;

export interface DynStepDefaultProps {
  spacing: LayoutSpacing;
  as: 'section';
  'data-testid': string;
}

export const DYN_STEP_DEFAULT_PROPS: DynStepDefaultProps = {
  spacing: 'md',
  as: 'section',
  'data-testid': 'dyn-step',
};
