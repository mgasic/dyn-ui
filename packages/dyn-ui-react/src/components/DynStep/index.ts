export { DynStep } from './DynStep';
export type {
  DynStepProps,
  DynStepRef,
  DynStepSpacingValue,
  DynStepOwnProps,
  DynStepBaseProps,
} from './DynStep.types';
