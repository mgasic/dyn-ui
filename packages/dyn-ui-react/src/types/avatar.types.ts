export type AvatarSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl';
export type AvatarLoading = 'eager' | 'lazy';

export interface DynAvatarProps {
  src?: string;
  size?: AvatarSize;
  loading?: AvatarLoading;
  alt?: string;
  initials?: string;
  className?: string;
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
}

export const AVATAR_SIZES = {
  xs: 24,
  sm: 32,
  md: 64,
  lg: 96,
  xl: 144
} as const;
