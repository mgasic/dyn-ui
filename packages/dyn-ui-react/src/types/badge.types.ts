export type {
  DynBadgeProps,
  DynBadgeRef,
  DynBadgeVariant,
  DynBadgeColor,
  DynBadgePosition,
  ComponentSize as DynBadgeSize,
  AccessibilityProps as DynBadgeAccessibilityProps,
  DYN_BADGE_COLORS
} from '../components/DynBadge/DynBadge.types';
