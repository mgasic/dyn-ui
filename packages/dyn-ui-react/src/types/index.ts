/**
 * Type exports for DYN UI React
 * Updated to include Layout Components types - SCOPE 7
 * Updated to include standardized BaseComponentProps - TYPE SYSTEM IMPROVEMENT
 */

// Base component props - standardized across all components
export type {
  BaseComponentProps,
  VariantProps,
  SizeProps
} from './theme';

// Theme types
export type { ThemeName, ThemeConfig, ThemeContextValue, ColorVariant, Size } from './theme';

// Display Components types - SCOPE 5
export type {
  DynBadgeProps,
  DynBadgeRef,
  DynBadgeVariant,
  DynBadgeColor,
  DynBadgePosition,
  DynBadgeSize,
  DynBadgeAccessibilityProps
} from './badge.types';
export type { DynAvatarProps, AvatarSize, AvatarLoading } from './avatar.types';
export type { DynIconProps, IconDictionary, ProcessedIcon } from './icon.types';
export type { DynLabelProps } from './label.types';

// Form Components types - SCOPE 6
export type {
  DynInputProps,
  DynSelectProps,
  DynDatePickerProps,
  DynFieldContainerProps,
  ValidationRule,
  DynFieldRef,
  DynFieldBase,
  SelectOption
} from './field.types';

export type {
  DynCheckboxProps,
  DynCheckboxRef,
  DynCheckboxSize,
} from '../components/DynCheckbox/DynCheckbox.types';
export {
  DYN_CHECKBOX_DEFAULT_PROPS,
} from '../components/DynCheckbox/DynCheckbox.types';

// Layout Components types - SCOPE 7
export type {
  DynContainerProps,
  DynContainerOwnProps,
  DynContainerBackground,
  DynPageProps,
  DynPageBreadcrumb,
  DynPageAction,
  LayoutSize,
  LayoutSpacing,
  LayoutDirection,
  LayoutAlignment,
  LayoutJustify
} from './layout.types';

export type {
  DynGridProps,
  DynGridColumn,
  DynGridPagination,
  DynGridSelectable,
  DynGridSortDirection,
} from '../components/DynGrid/DynGrid.types';
export { DYN_GRID_DEFAULT_PROPS } from '../components/DynGrid/DynGrid.types';

// Constants
export { AVATAR_SIZES } from './avatar.types';
