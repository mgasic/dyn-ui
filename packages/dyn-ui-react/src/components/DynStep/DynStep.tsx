import React, { forwardRef } from 'react';
import { cn } from '../../utils/classNames';
import type { DynStepProps, DynStepRef, DynStepSpacingValue } from './DynStep.types';
import { DYN_STEP_DEFAULT_PROPS } from './DynStep.types';
import styles from './DynStep.module.css';

const SPACING_TOKEN_MAP = {
  none: '0',
  xs: 'var(--dyn-spacing-xs, 0.25rem)',
  sm: 'var(--dyn-spacing-sm, 0.5rem)',
  md: 'var(--dyn-spacing-md, 0.75rem)',
  lg: 'var(--dyn-spacing-lg, 1rem)',
  xl: 'var(--dyn-spacing-xl, 1.5rem)',
} as const;

type LayoutSpacingKey = keyof typeof SPACING_TOKEN_MAP;

const isLayoutSpacingToken = (value: DynStepSpacingValue | undefined): value is LayoutSpacingKey => {
  if (typeof value !== 'string') return false;
  return value in SPACING_TOKEN_MAP;
};

const resolveSpacingValue = (value: DynStepSpacingValue | undefined) => {
  if (value === undefined) return undefined;
  if (typeof value === 'number') {
    return value === 0 ? '0' : `${value}px`;
  }
  if (isLayoutSpacingToken(value)) {
    return SPACING_TOKEN_MAP[value];
  }
  return value;
};

export const DynStep = forwardRef(function DynStep<
  E extends React.ElementType = typeof DYN_STEP_DEFAULT_PROPS.as
>(
  {
    as,
    spacing = DYN_STEP_DEFAULT_PROPS.spacing,
    padding,
    className,
    style,
    children,
    'data-testid': dataTestId = DYN_STEP_DEFAULT_PROPS['data-testid'],
    ...rest
  }: DynStepProps<E>,
  ref: DynStepRef<E>
) {
  const Component = (as ?? DYN_STEP_DEFAULT_PROPS.as) as React.ElementType;

  const gapValue = resolveSpacingValue(spacing ?? DYN_STEP_DEFAULT_PROPS.spacing);
  const paddingValue = resolveSpacingValue(padding);

  const cssVariables: React.CSSProperties = {
    ...(gapValue ? { ['--dyn-step-gap' as const]: gapValue } : {}),
    ...(paddingValue !== undefined ? { ['--dyn-step-padding' as const]: paddingValue } : {}),
  };

  return (
    <Component
      ref={ref as any}
      className={cn(styles.root, className)}
      data-testid={dataTestId}
      style={{ ...cssVariables, ...style }}
      {...rest}
    >
      {children}
    </Component>
  );
});

DynStep.displayName = 'DynStep';

export default DynStep;
