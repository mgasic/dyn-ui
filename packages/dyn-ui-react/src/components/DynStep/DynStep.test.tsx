import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { DynStep } from './DynStep';
import { DYN_STEP_DEFAULT_PROPS } from './DynStep.types';

describe('DynStep', () => {
  it('exports a React component', () => {
    expect(DynStep).toBeDefined();
    expect((DynStep as any).displayName).toBe('DynStep');
  });

  it('renders with default props as a section', () => {
    render(<DynStep>Content</DynStep>);
    const element = screen.getByTestId(DYN_STEP_DEFAULT_PROPS['data-testid']);

    expect(element).toBeInTheDocument();
    expect(element.tagName).toBe('SECTION');
  });

  it('supports polymorphic rendering via the `as` prop', () => {
    render(
      <DynStep as="article" data-testid="polymorphic-step">
        Article content
      </DynStep>
    );

    const element = screen.getByTestId('polymorphic-step');
    expect(element.tagName).toBe('ARTICLE');
  });

  it('applies spacing tokens as CSS variables', () => {
    render(
      <DynStep spacing="lg" padding="sm" data-testid="spaced-step">
        Spaced content
      </DynStep>
    );

    const element = screen.getByTestId('spaced-step');

    expect(element).toHaveStyle({
      '--dyn-step-gap': 'var(--dyn-spacing-lg, 1rem)',
      '--dyn-step-padding': 'var(--dyn-spacing-sm, 0.5rem)',
    });
  });

  it('supports custom spacing values', () => {
    render(
      <DynStep spacing={12} padding="24px" data-testid="custom-spacing">
        Custom spacing
      </DynStep>
    );

    const element = screen.getByTestId('custom-spacing');

    expect(element).toHaveStyle({
      '--dyn-step-gap': '12px',
      '--dyn-step-padding': '24px',
    });
  });

  it('matches snapshot with spacing props', () => {
    const { asFragment } = render(
      <DynStep spacing="xl" padding="lg" data-testid="snapshot-step">
        <header>
          <h2>Review &amp; Submit</h2>
          <p>Confirm your entries before finishing.</p>
        </header>
        <div>
          <p>All fields look good. Click finish to complete the wizard.</p>
        </div>
      </DynStep>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
