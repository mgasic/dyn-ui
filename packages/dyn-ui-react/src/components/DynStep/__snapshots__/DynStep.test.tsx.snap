// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DynStep > matches snapshot with spacing props 1`] = `
<DocumentFragment>
  <section
    class="_root_27c983"
    data-testid="snapshot-step"
    style="--dyn-step-gap: var(--dyn-spacing-xl, 1.5rem); --dyn-step-padding: var(--dyn-spacing-lg, 1rem);"
  >
    <header>
      <h2>
        Review & Submit
      </h2>
      <p>
        Confirm your entries before finishing.
      </p>
    </header>
    <div>
      <p>
        All fields look good. Click finish to complete the wizard.
      </p>
    </div>
  </section>
</DocumentFragment>
`;
